---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

# TidyJaccard

<!-- badges: start -->
[![Lifecycle: experimental](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://lifecycle.r-lib.org/articles/stages.html#experimental)
[![R-CMD-check](https://github.com/selkamand/tidyjaccard/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/selkamand/tidyjaccard/actions/workflows/R-CMD-check.yaml)
[![CRAN status](https://www.r-pkg.org/badges/version/tidyjaccard)](https://CRAN.R-project.org/package=tidyjaccard)
[![Codecov test coverage](https://codecov.io/gh/selkamand/tidyjaccard/branch/master/graph/badge.svg)](https://app.codecov.io/gh/selkamand/tidyjaccard?branch=master)
<!-- badges: end -->

Effortless all vs all Jaccard similarity from tidy dataframes.

## Philosophy

For TidyJaccard to work seamlessly, ensure that your dataframe includes two columns:

1. **sample_identifier**: This column should contain unique identifiers for each sample in your dataset.

2. **set**: Each row in this column defines an element within the set corresponding to the respective sample identifier. This column plays a pivotal role in calculating Jaccard similarity.

Names of columns doesn't matter. 



## Installation

You can install the development version of tidyjaccard like so:

``` r
remotes::install_github('selkamand/tidyjaccard')
```


## Getting Started


Here's an example [dataset](https://github.com/selkamand/tidyjaccard/blob/master/inst/smith_and_johnson.csv) describing physical traits of two families

- Family 1: Smith (5 members)

- Family 2: Johnson (3 members)

*Research Question:* Which people in the dataset are most physically similar to one another? Do they tend to come from the same family?

For each member of the family, we've recorded physical traits. Each row in our dataset describes one trait present in one individual.

Lets look at the data:

```{r}
library(tidyjaccard)

# Inbuilt Dataset Describing Smith and Johnson family members
head(family)
```

### Compute Jaccard Similarity

```{r}
# Compute similarity
family |>
  tidy_pairwise_jaccard_similarity(col_sample = "name", col_trait = "trait") # Compute Jaccard Similarity
```


### Convert to distance matrix

We can convert the tidy similarity to a distance matrix for use with many functions

```{r}
family |>
  tidy_pairwise_jaccard_similarity(col_sample = "name", col_trait = "trait") |> # Compute Jaccard Similarity
  tidy_toggle_simdist() |> # Convert similarity to distance
  tidy_to_matrix() |> # Convert to matrix
  matrix_to_dist_class() # Convert matrix to 'dist' class (for compatibility with hclust)
```

### Heirarchical Clustering and Visualisation

Heirarchical clustering is as simple as running `hclust()` on the distance matrix. Then just pipe to `visualise_dendogram_ggraph()` to visualise the relationships

```{r}
family |>
  tidy_pairwise_jaccard_similarity(col_sample = "name", col_trait = "trait") |> # Compute Jaccard Similarity
  tidy_toggle_simdist() |> # Convert similarity to distance
  tidy_to_matrix() |> # Convert to matrix
  matrix_to_dist_class() |> # Convert matrix to 'dist' class (for compatibility with hclust)
  hclust(method = "complete") |> # Do heirarchical clustering
  visualise_dendrogram_ggraph(draw_labels = TRUE) # Visualise the dendrogram
```


###  Customising Visualisation


Finally lets map some metadata about each family member to different visual properties. 

Requirements: a dataframe with 1 row per leaf node. The first column will be our describe the label of each leaf node (e.g. Mia Johnson, Liam Johnson etc). This dataframe should also include columns we can map to colour, shape, and tooltips (for when we make the plot interactive).

TidyJaccard come pre-packaged with an example dataset `family_member_annotations`

```{r}
print(family_member_annotations) # leaf node annotation dataset
```


```{r}
gg_dendogram <- family |>
  tidy_pairwise_jaccard_similarity(col_sample = "name", col_trait = "trait") |> # Compute Jaccard Similarity
  tidy_toggle_simdist() |> # Convert similarity to distance
  tidy_to_matrix() |> # Convert to matrix
  matrix_to_dist_class() |> # Convert matrix to 'dist' class (for compatibility with hclust)
  hclust(method = "complete") |> # Do heirarchical clustering
  visualise_dendrogram_ggraph( # Visualise the dendrogram
    annotations = family_member_annotations, # Dataframe with 1 row per labelled node
    col_label = name, 
    col_colour = family, 
    col_shape = family,
    col_tooltip = tooltip
    ) 

gg_dendogram
```

### Adding Interactivity 

Make the dendrogram interactive using `visualisation_make_interactive()`

```{r, eval = FALSE}
visualisation_make_interactive(gg_dendogram)
```


### Working in Parallel

If `tidy_pairwise_jaccard_similarity()` is too slow for your liking, get an easy speed up by using multiple threads.

This functionality is powered by the `futureverse` 

```{r}
# Turn on multithreading
library(future)
threads = parallelly::availableCores(omit = 1) # How many threads to use. This will use all available cores except 1
plan(multisession, workers = threads)

# Compute jaccard similarity
family |>
  tidy_pairwise_jaccard_similarity(col_sample = "name", col_trait = "trait") # Compute Jaccard Similarity
```



## Examples of Other Applications

RNAseq data

- col_sample = sample identifier
- col_trait = overexpressed genes


Microbiome data

- col_sample = sample identifier
- col_trait = microbes present
